src/Dhall/Import.hs

```diff
@@ -662,21 +667,35 @@ exprFromPath m (Path {..}) = case pathType of
     This also returns the true final path (i.e. explicit "/@" at the end for
     directories)
 -}
-loadDynamic :: Path -> StateT Status IO (Expr Src Path)
-loadDynamic p = do
+loadDynamic :: forall m . MonadCatch m => (Path -> m (Expr Src Path))
+    -> Path -> StateT Status m (Expr Src Path)
+loadDynamic from_path p = do
     paths <- zoom stack State.get
 
-    let handler :: SomeException -> IO (Expr Src Path)
-        handler e = throwIO (Imported (p:paths) e)
-    m <- needManager
-    liftIO (exprFromPath m (canonicalizePath (p:paths)) `catch` handler)
+    let handler :: SomeException -> m (Expr Src Path)
+        handler e = throwM (Imported (p:paths) e)
 
--- | Load a `Path` as a \"static\" expression (with all imports resolved)
-loadStatic :: Path -> StateT Status IO (Expr Src X)
-loadStatic = loadStaticWith Dhall.Context.empty
+    lift (from_path (canonicalizePath (p:paths)) `catch` handler)
 
-loadStaticWith :: Dhall.Context.Context (Expr Src X) -> Path -> StateT Status IO (Expr Src X)
-loadStaticWith ctx path = do
+loadStaticIO :: Dhall.Context.Context (Expr Src X) -> Path -> StateT Status IO (Expr Src X)
+loadStaticIO ctx path = do
+    m <- needManager
+    loadStaticWith (exprFromPath m) ctx path
+
+-- | Resolve all imports within an expression using `Path` resolving callback
+-- in arbitrary `MonadCatch` monad.
+load' :: MonadCatch m => (Path -> m (Expr Src Path)) -> Expr Src Path -> m (Expr Src X)
+load' from_path = loadWith' from_path (Dhall.Context.empty)
+
+-- | Resolve all imports within an expression using a custom typing context and Path
+-- resolving callback in arbitrary `MonadCatch` monad.
+loadWith' :: MonadCatch m => (Path -> m (Expr Src Path))
```

Comment:
> For these generalized functions that run in MonadCatch, I think you can generalize their types to replace the Path with a type variable.

It looks like significant parts of `Dhall.Import` deal with `Path` data type itself. E.g. see `canonicalizePath`, local path handling in `loadStaticWith` etc. Doing `loadWith` with generic
type would require having this functionality in some sort of callback...

> ... but with the current pull-request that wouldn't work because Big will not parse as a Path, so the user would have to provide a path like ./Big instead, which might not be what you want. However, if you were to generalize loadWith' to work with things other than Path (with matching changes to the parser) then you can support your original example

I am indeed doing the dirty `./Big` hack currently.

To fix it properly, I would suggest changing `Path` to be `Path a` and adding something like `Custom a` constructor. Then we can leverage `exprA` parsing to parse some kind of `custom:Big` special `Path`. Also the `custom` prefix can be something which the user defines.

IMHO, this kind of change is better done independently from the current pull request.